compiler/rustc_codegen_ssa/src/back/link.rs

```diff
@@ -635,57 +653,33 @@ fn escape_stdout_stderr_string(s: &[u8]) -> String {
-
-/// Invoke `llvm-dwp` (shipped alongside rustc) to link `dwo` files from Split DWARF into a `dwp`
-/// file.
-fn link_dwarf_object<'a>(sess: &'a Session, executable_out_filename: &Path) {
-    info!("preparing dwp to {}.dwp", executable_out_filename.to_str().unwrap());
-
+/// Invoke `thorin` (rust implementation of a dwarf packaging utility) to link DWARF objects into a
+/// DWARF package.
+fn link_dwarf_object<'a>(
+    sess: &'a Session,
+    codegen_results: &CodegenResults,
+    executable_out_filename: &Path,
+) {
+    let input_rlibs = codegen_results
+        .crate_info
+        .used_crate_source
+        .values()
+        .filter_map(|csource| csource.rlib.clone())
+        .map(|(path, _)| path)
+        .collect::<Vec<_>>();
     let dwp_out_filename = executable_out_filename.with_extension("dwp");
-    let mut cmd = Command::new(LLVM_DWP_EXECUTABLE);
-    cmd.arg("-e");
-    cmd.arg(executable_out_filename);
-    cmd.arg("-o");
-    cmd.arg(&dwp_out_filename);
-
-    let mut new_path = sess.get_tools_search_paths(false);
-    if let Some(path) = env::var_os("PATH") {
-        new_path.extend(env::split_paths(&path));
-    }
-    let new_path = env::join_paths(new_path).unwrap();
-    cmd.env("PATH", new_path);
+    let executable_out_filename = executable_out_filename.to_path_buf();
 
-    info!("{:?}", &cmd);
-    match sess.time("run_dwp", || cmd.output()) {
-        Ok(prog) if !prog.status.success() => {
-            sess.struct_err(&format!(
-                "linking dwarf objects with `{}` failed: {}",
-                LLVM_DWP_EXECUTABLE, prog.status
-            ))
-            .note(&format!("{:?}", &cmd))
-            .note(&escape_stdout_stderr_string(&prog.stdout))
-            .note(&escape_stdout_stderr_string(&prog.stderr))
-            .emit();
-            info!("linker stderr:\n{}", escape_stdout_stderr_string(&prog.stderr));
-            info!("linker stdout:\n{}", escape_stdout_stderr_string(&prog.stdout));
-        }
-        Ok(_) => {}
+    debug!(?input_rlibs, ?dwp_out_filename, ?executable_out_filename);
+    match sess.time("run_thorin", || {
+        thorin::package(input_rlibs, Some(vec![executable_out_filename]), dwp_out_filename)
+    }) {
+        Ok(()) => {}
         Err(e) => {
-            let dwp_not_found = e.kind() == io::ErrorKind::NotFound;
-            let mut err = if dwp_not_found {
-                sess.struct_err(&format!("linker `{}` not found", LLVM_DWP_EXECUTABLE))
-            } else {
-                sess.struct_err(&format!("could not exec the linker `{}`", LLVM_DWP_EXECUTABLE))
-            };
-
-            err.note(&e.to_string());
-
-            if !dwp_not_found {
-                err.note(&format!("{:?}", &cmd));
-            }
-
-            err.emit();
+            sess.struct_err("linking dwarf objects with thorin failed")
+                .note(&format!("{:?}", e))
+                .emit();
+            info!("linker err:\n{}", format!("{:?}", e));
```

Comment:
I don't think there is a need for this given that there is already an error with the exact same information.

Reply:
Fixed!